src/lazy/expanded/struct.rs

```diff
@@ -580,34 +654,240 @@ impl<'top, D: Decoder> ExpandedStructIterator<'top, D> {
+                                // back to reading from the source.
+                                *state = ReadingFieldFromSource;
+                            }
+                        }
+                    }
+                }
             }
-        };
-        let struct_ = match expanded_value.read()? {
-            ExpandedValueRef::Struct(s) => s,
-            other => {
-                return IonResult::decoding_error(format!(
-                    "macros in field name position must produce structs; '{:?}' produced: {:?}",
-                    invocation, other
-                ))
+        }
+    }
+
+    #[cfg(test)]
+    mod tests {
+        use super::*;
+        use crate::{v1_1, Element, MacroExprKind, Reader};
+
+        #[test]
+        fn field_kinds() -> IonResult<()> {
+            let source = r#"
+                $ion_1_1
+                (:add_macros
+                    (macro three_values ()
+                        (.values 1 2 3)
+                    )
+                    (macro three_structs ()
+                        (.values {dog: 1} {cat: 2} {mouse: 3})
+                    )
+                )
+                {
+                    foo: 0,
+                    bar: (:three_values),
+                    (:three_structs),
+                    quux: true,
+                }
+            "#;
+            let mut reader = Reader::new(v1_1::Text, source)?;
+            let struct_ = reader.expect_next()?.read()?.expect_struct()?;
+            let fields = &mut struct_.expanded_struct.field_exprs();
+
+            fn expect_name_value<'top, D: Decoder>(
+                fields: &mut impl Iterator<Item = IonResult<FieldExpr<'top, D>>>,
+                expected_name: &str,
+                expected_value: impl Into<Element>,
+            ) -> IonResult<()> {
+                let field = fields.next().unwrap()?;
+                let expected_value = expected_value.into();
+                assert!(
+                    matches!(
+                        field,
+                        FieldExpr::NameValue(name, value)
+                            if name.read()?.text() == Some(expected_name)
+                            && Element::try_from(value.read_resolved()?)? == expected_value,
+                    ),
+                    "{field:?} did not match name={expected_name:?}, value={expected_value:?}"
+                );
+                Ok(())
             }
-        };
-        let iter: &'top mut ExpandedStructIterator<'top, D> = struct_.bump_iter();
-        *state = ExpandedStructIteratorState::InliningAStruct(iter);
-        Ok(())
+
+            expect_name_value(fields, "foo", 0)?;
+            assert!(matches!(
+                fields.next().unwrap()?,
+                FieldExpr::NameMacro(name, invocation)
+                    if name.read()?.text() == Some("bar") && matches!(invocation.kind(), MacroExprKind::EExp(eexp) if eexp.invoked_macro.name() == Some("three_values"))
+            ));
```

Comment:
On the one hand, it's cool that we can do this in Rust. On the other hand... 😨

I wish there was a better syntax for expressing a set of constraints like this.